Firmware 2.0 compatibility

MaxandreOgeret opened the pull request:
**Big thanks to Ouster for lending me a sensor to continue this development !**

fixes #66 

Brought client directory from https://github.com/ouster-lidar/ouster_example
This PR still needs a lot of cleaning.

**Include**
 - ros2_ouster
   -  client
     - client.h
     - types.h
     - version.h
     - impl
       - netcompat.h
       - parsing.h

**src**
 - client
   - client.cpp
   - netcompat.cpp
   - types.cpp

I followed the idea in this comment : https://github.com/ros-drivers/ros2_ouster_drivers/issues/63#issuecomment-742027046. But a lot of changes had to be made nonetheless. For example `ClientState` became `client_state` and `Metadata` became `client_info`.  Also the `client_state` can now have several values that can be determined with a bitwise operation. (See `toString` in `conversions.h`)

It compiles BUT it's not functionnal. It doesn't publish anything at the moment. I reaches the `process` function in the processor, for example `pointcloud_processor.hpp` but never actually publishes anything. It seems that the `_batch_and_publish` function fails somehow and that we never reach it. 

And at this point I'm kinda stuck, I don't know what is wrong. Could someone have a look at it ? @SteveMacenski ?

Also, should the class `LidarScan` be brought to the repo ? Because it seems that how pointclouds are currently computed is deprecated somehow. In the ouster repo they compute scans first and then pointclouds from scans.
The files `OS1_packet.hpp`, `OS1_sensor.hpp`, `OS1_util.hpp`

If we are required to compute the pointclouds from scans how do you think we should proceed because at the moment we compute it from the `packet_data`. But maybe there's still a way to do it.

Also, In the ouster repo there is no mention of OS1 sensor, does it still make sense to have a `OS1` class and not something like `sensor` ?

Thanks for your help, I will continue to work on this when I have time again.

MaxandreOgeret committed:
initial commit firmware 2.0 compatibility

anderwm commented:
Echo that this is more complicated than meets the eye. There are types in the V2 example that conflict with types from V1x and don't necessarily do the same thing. With only a shallow understanding of the code here, it is hard to piece it all back together as the writer intended.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -56,58 +56,68 @@ static const bool IS_BIGENDIAN = []()
+  if (state & ouster::sensor::client_state::LIDAR_DATA) {
+    output << "lidar data ";
+  }
+  if (output.str() == "") {
```
```suggestion
  if (output.empty()) {
```

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -56,58 +56,68 @@ static const bool IS_BIGENDIAN = []()
+  msg.hostname = mdata.name;
   msg.lidar_mode = mdata.mode;
-  msg.timestamp_mode = mdata.timestamp_mode;
+//  msg.timestamp_mode = mdata.timestamp_mode; //todo: Find how to get timestamp_mode
```
This is why we had a custom metadata class in `ros2_ouster` vs the client header version so that we could store additional information like this.

MaxandreOgeret replied:
So the metadata in `ros2_ouster` is the same as the client with some additions ? It would make sense to make the `ros2_ouster` metadata align with the new metadata then.

But what about the client functions that use the client metadata like this [one](https://github.com/ouster-lidar/ouster_example/blob/master/ouster_client/src/types.cpp#L112)  ? Should I re-implement it in `ros2_ouster` to make use of the `ros2_ouster` metadata ? (It seems like a good idea. I should create `metadata.cpp` and implement them there).

SteveMacenski replied:
> So the metadata in ros2_ouster is the same as the client with some additions ? It would make sense to make the ros2_ouster metadata align with the new metadata then.

Yes, exactly! Plus whatever other information is just convenient (like the angle vectors) that stay the same over time.

> But what about the client functions that use the client metadata like this one ? Should I re-implement it in ros2_ouster to make use of the ros2_ouster metadata ? (It seems like a good idea. I should create metadata.cpp and implement them there).

I'm not sure, it depends if that's necessary to store / tell users.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -56,58 +56,68 @@ static const bool IS_BIGENDIAN = []()
   msg.beam_altitude_angles = mdata.beam_altitude_angles;
-  msg.imu_to_sensor_transform = mdata.imu_to_sensor_transform;
-  msg.lidar_to_sensor_transform = mdata.lidar_to_sensor_transform;
+  msg.imu_to_sensor_transform = std::vector<double>(mdata.imu_to_sensor_transform.data(), mdata.imu_to_sensor_transform.data() + mdata.imu_to_sensor_transform.rows() * mdata.imu_to_sensor_transform.cols());
```
I'm not sure I'm following the transform population here - but it shouldn't stop data from coming through so we can ignore this for right now.

MaxandreOgeret replied:
Here I transform the eigen matrix to a vector. Any better idea how to do that ?

SteveMacenski replied:
Store it as a vector or write into a method

MaxandreOgeret replied:
Should be solved now. I created a function to convert the mat4d into the vector.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -56,58 +56,68 @@ static const bool IS_BIGENDIAN = []()
-  tf.setOrigin({mat[3] / 1e3, mat[7] / 1e3, mat[11] / 1e3});
-  tf.setBasis(
-    {mat[0], mat[1], mat[2], mat[4], mat[5], mat[6], mat[8], mat[9], mat[10]});
+  tf.setOrigin({*(mat.data()+3) / 1e3, *(mat.data()+7) / 1e3, *(mat.data()+11) / 1e3});
```
This is a strange way to access an Eigen Matrix - why not use the normal `mat(4)` type accessors? https://eigen.tuxfamily.org/dox/group__TutorialMatrixClass.html

MaxandreOgeret replied:
`mat(4)` ? You mean something like  `mat(i,j)` ?

SteveMacenski replied:
However it looks, yeah. I think this is a 1xd so I think its just `(x)` but I could be wrong

MaxandreOgeret replied:
You are right, mat(x) is fine ! Thx !

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -241,23 +251,22 @@ inline sensor_msgs::msg::LaserScan toMsg(
 
   double resolution, rate;
-  if (mdata.mode == "512x10") {
+  if (mdata.mode == ouster::sensor::lidar_mode::MODE_512x10) {
```
Are there different / more modes now for V2?

SteveMacenski commented on ros2_ouster/src/OS1/OS1_sensor.cpp:
```diff
@@ -43,41 +40,47 @@ void OS1Sensor::reset(const ros2_ouster::Configuration & config)
             std::string("Failed to create connection to lidar."));
   }
+
+  _pf = &ouster::sensor::get_format(getMetadata());
```
Are you checking that this is working? I could see this being why you're not getting data out if this size is wrong / empty / not being retrieved so the packet size is garbage.

MaxandreOgeret replied:
`mat(4)` ? You mean something like  `mat(i,j)` ?

MaxandreOgeret replied:
What would be the best way to test that for you ? Or how would you create the `_pf` class variable ?

The struct is defined [here](https://github.com/ouster-lidar/ouster_example/blob/master/ouster_client/include/ouster/types.h#L174). And it has const elements that I would have to initialize no ?

So using a pointer seems to be the best way no ?

Here I output some values from the struct : 
```
[ouster_driver-1] lidar_packet_size 24896
[ouster_driver-1] imu_packet_size 48
[ouster_driver-1] columns_per_packet 16
[ouster_driver-1] pixels_per_column 128
[ouster_driver-1] encoder_ticks_per_rev 90112
```

It's hard to judge but it seems like this works.

SteveMacenski replied:
That seems reasonable... so when you see the packets in the processors, are those looking OK (e.g. the XYZ intensity, etc values look OK if you print them)?

MaxandreOgeret replied:
Yes everything looked fine and the driver works now.

SteveMacenski commented on ros2_ouster/src/OS1/OS1_sensor.cpp:
```diff
@@ -87,34 +90,29 @@ ros2_ouster::ClientState OS1Sensor::get()
-    default:
-      return nullptr;
+  // todo : Manage case when we receive lidar + imu data
+  if (state & ouster::sensor::client_state::LIDAR_DATA) {
```
Are these things bitwise embedded now?

SteveMacenski replied:
Maybe also why not getting data if these aren't called?

MaxandreOgeret replied:
Yes they are, example how it's hone [here](https://github.com/ouster-lidar/ouster_example/blob/5c00aa97ee8244072ed9bf14870facf57d4aab98/ouster_ros/src/os_node.cpp#L84)

I've checked and the packets are identified well. How do we handle the case when a package has lidar + imu data ?

SteveMacenski replied:
It doesn't happen

SteveMacenski replied:
but I suppose if it did, since that's the reference code I assume this handles it- but unless something's really changed on firmware, the packets are separated

MaxandreOgeret replied:
As is turns out and even though packages are separated it is now possible to have lidar packets and imu packets available when we reach `processData()`. And it happened several times in my tests.
Also now the check for `LIDAR_DATA` or `IMU_DATA` is bitwise so it's possible to have both TRUE.
I changed the logic to get the packets:

- Reaplaced `readPacket` with `readLidarPacket` and `readImuPacket`
- In `processData()` they are now processed separately in two different loops.

SteveMacenski commented on ros2_ouster/src/OS1/OS1_sensor.cpp:
```diff
@@ -87,34 +90,29 @@ ros2_ouster::ClientState OS1Sensor::get()
-    return {"UNKNOWN", "UNKNOWN", "UNNKOWN", "UNNKOWN", "UNKNOWN",
-      {}, {}, {}, {}, 7503, 7502};
-  }
+//  if (_ouster_client) {
```
Readd this - its for your own protection in case things aren't set up correctly

MaxandreOgeret replied:
I will when I will make use of the `ros2_ouster` metadata.

MaxandreOgeret replied:
To take care of this I implemented the ouster provided function `populate_missing_metadata_defaults` in `metadata.hpp`.

SteveMacenski commented:
@MaxandreOgeret please explain where you've tracked the data down to (e.g. where's the last place in the code you know its right or where's the first part of the code you know something's not triggering). 

@anderwm do you see something wrong here or a function we're misusing? 

> Also, should the class LidarScan be brought to the repo ?

I don't understand the question, please provide links :smile: . 

> I reaches the process function in the processor, for example pointcloud_processor.hpp but never actually publishes anything.

Did you check that the packets its given are "real"? I'd try just publishing data each iteration to the terminal and soft sanity check the ranges / values look "real". The batch function didn't change it doesn't look so the only thing that could have changed are the inputs or how the inputs are formatted. 

> Also, In the ouster repo there is no mention of OS1 sensor, does it still make sense to have a OS1 class and not something like sensor ?

We can rename it once we get it working. I'd rather focus on functionality and then go back over and make sure that we make use of all the right things and named right.

anderwm commented:
> Also, should the class LidarScan be brought to the repo ? Because it seems that how pointclouds are currently computed is deprecated somehow. In the ouster repo they compute scans first and then pointclouds from scans.
The files OS1_packet.hpp, OS1_sensor.hpp, OS1_util.hpp
If we are required to compute the pointclouds from scans how do you think we should proceed because at the moment we compute it from the packet_data. But maybe there's still a way to do it.

@MaxandreOgeret 's point above is kind of what I'm struggling to understand as well. I freely admit I am nowhere close to understanding all you have done here. However, I can't see how the functionality exposed in [their LidarScan](https://github.com/ouster-lidar/ouster_example/blob/master/ouster_client/include/ouster/lidar_scan.h) class can be replicated. Or even if it can be, if it would be prudent to do so. Particularly, the way ScanBatcher makes a LidarScan from a scan of data seems like a significant shift from the paradigm they used before. On the other hand, bringing in LidarScan opens up a whole different sack of ferrets (mostly about what gets stripped out/ where it hooks into the ros2_ouster code).

It could also be that I don't understand the code well enough to see the easy way out...

SteveMacenski commented:
IMO anything in this directory https://github.com/ros-drivers/ros2_ouster_drivers/tree/foxy-devel/ros2_ouster/include/ros2_ouster/OS1 is fair game to nuke. We need to have some time of abstraction (https://github.com/ros-drivers/ros2_ouster_drivers/blob/foxy-devel/ros2_ouster/include/ros2_ouster/OS1/OS1_sensor.hpp) for the specific sensor so that we separate the driver from the sensor and I want to make sure we continue to follow the processor (https://github.com/ros-drivers/ros2_ouster_drivers/tree/foxy-devel/ros2_ouster/include/ros2_ouster/OS1/processors) design for throwing a packet from the sensor at these processors so that you can create different data structures (XYZ point cloud, 2D laser scan, IMU, XYZI Pointcloud, etc).

How that data is buffered and what functions we use (e.g. remove the batch iterator lambdas) is totally fair game to remove if the V2 code demands it. The ROS 1 driver changes how it batches packets into complete pointclouds to return (looks like it does https://github.com/ouster-lidar/ouster_example/blob/master/ouster_ros/src/os_cloud_node.cpp#L60) I'm not married to keeping it in the V1.x styling at all.

MaxandreOgeret commented:
Thanks both of you for this analysis. I need to wrap my head around all this and continue working on it. I should have some time this weekend.

> please explain where you've tracked the data down to (e.g. where's the last place in the code you know its right or where's the first part of the code you know something's not triggering).

For me the connection with the lidar is OK, polling is OK, reading the packets is OK.
When I output the `_lidar_packet` as hex I get data so it should be fine.

I manage to get into the `process` function of the processor but then nothing happens, it seems that we never get into the `_batch_and_publish` lambda. And the lambda is very dense and I'm not sure what is happening there.

For example I cannot output anything with `cout` in the lambda.

> Did you check that the packets its given are "real"?

They seem to be ! Here is the output in hex of one packet : https://pastebin.com/WwR39rEV

---

It seems like it would be a good idea imitate the way packets are batched in the ROS1 client in the processors. BTW, what was the idea behind using lambdas ?

SteveMacenski commented:
> I manage to get into the process function of the processor but then nothing happens, it seems that we never get into the _batch_and_publish lambda. And the lambda is very dense and I'm not sure what is happening there.

So maybe part of the disconnect here, that lambda is not mine, that's from the V1.x client from ouster itself. So it might be that things look so different now that that just doesn't work anymore. you can just remove that lambda and use whatever the new batching system is for the V2.x client instead. 

> It seems like it would be a good idea imitate the way packets are batched in the ROS1 client in the processors. But as you can see here they are computing scans and then later pointclouds.

Sure but look at that function: https://github.com/ouster-lidar/ouster_example/blob/9352730127c3e2bc8ce1e778be4ae8b0a3b40f3c/ouster_ros/src/ros.cpp#L67, its trivial. All it does is iterate over the data to populate the pointcloud. That can just be removed if the pointcloud processor is doing that in parallel. 

I haven't looked at the scan class that closely, but it seems like its just the laser scans from the sensor (e.g. the pointcloud) in an internal data structure that's easier to store into and then converted into a PC2 type for publication. It doesn't seem to me that it does anything worth keeping. It seems like the packets are "batched" to a full rotation "scan" which is then converted into a "pointcloud" https://github.com/ouster-lidar/ouster_example/blob/master/ouster_client/src/example.cpp. The only reason that I think the scan is useful is if it is used to convert into the depth / reflectivity / intensity images too, but I don't see that happening in the code.

MaxandreOgeret commented:
> you can just remove that lambda and use whatever the new batching system is for the V2.x client instead.

Yep, will do that.

> Maybe there's some word choice issue here (...)
> I think its just a full rotation of the 3D lidar accumulating the packets to then populate the pointcloud.

Indeed it seems so, things seem to be a bit clearer now, thanks for catching this.  
So we would need to 'skip' the scan. I will try to see what I can do.

SteveMacenski commented:
Got it, sounds good! The 2D scan thing is something I added to the ROS2 drivers (and more of the image topics) because alot of people just want the 2D laser scan as another topic for localization / mapping on embedded systems but use the full 3D pointcloud for collision avoidance. You could always extract it later, but its more efficient to do it at this point.

MOG-Work committed:
now use ros2 ouster metadata

MOG-Work committed:
added lidar_scan + brought pf to pc processor

MOG-Work committed:
added point.h from client + finished pointcloud processor.

MaxandreOgeret commented:
Today I got the first publishing from the sensor, it' s not working yet and I suspect the packages are wrong. I need to have a look at the `_pf ` variable again, as it might be the cause of the problems.

But I' m on the right path ! :smile: 

<details>

  <summary>Invalid pointcloud2 message : </summary> 

```
---
header:
  stamp:
    sec: 5269
    nanosec: 860104870
  frame_id: laser_data_frame
height: 128
width: 1024
fields:
- name: x
  offset: 0
  datatype: 7
  count: 1
- name: y
  offset: 4
  datatype: 7
  count: 1
- name: z
  offset: 8
  datatype: 7
  count: 1
- name: intensity
  offset: 16
  datatype: 7
  count: 1
- name: t
  offset: 20
  datatype: 6
  count: 1
- name: reflectivity
  offset: 24
  datatype: 4
  count: 1
- name: ring
  offset: 26
  datatype: 2
  count: 1
- name: ambient
  offset: 28
  datatype: 4
  count: 1
- name: range
  offset: 32
  datatype: 6
  count: 1
is_bigendian: false
point_step: 48
row_step: 49152
data:
- 0
- 0
- 0
- 128
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 128
- 63
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 128
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 128
- 63
- 0
- 0
- 0
- 0
- 90
- 69
- 220
- 3
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 128
- 63
- 0
- 0
- 0
- 0
- 136
- 71
- 34
- 204
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- 0
- '...'
is_dense: true
---

```
</details>

Commit:
solved "bad alloc error" when an OusterDriverException was thrown

MOG-Work committed:
removed packet format pointer + driver packet buffer now class variable + added metadata defaults

SteveMacenski commented:
>  I also got rid of the lambda and implemented a standard function. I think it' s way cleaner like that,

Awesome! Yeah that lambda was a real pain to work through myself when I ported that from the ROS 1 drivers. I think the PC2 processor is the easiest to start with (easiest to visually verify). Then IMU, laser scan, then the image processors. 

Unfortunately, I'm still not on V2 yet to be able to test any changes :(

MaxandreOgeret commented:
Thanks for your comment !
No problem I understand, I'm also not very fast myself. I should get a second sensor in a few weeks dedicated only to development so it will go faster starting from there.
I tested today and I have a problem somewhere between getting the packet data and turning it into a pc2. It's not fast but I'm on it ! :)

SteveMacenski commented:
No worries, you're on it more than I am right now, so all the credit to you :laughing:. Q1 of this year hasn't gone at all how I expected it to - sorry if it makes you feel better, this is by no means the only thing I've not been as active on as I would like due to other things that have exploding deadlines on my side!

MOG-Work committed:
simplified pc processor & pc2 conversion for testing purposes

MaxandreOgeret commented:
I have some questions @SteveMacenski.

Sorry for the flow development time, it's hard to test what I do as I have to run to the lab after working hours as I am remote working because of the covid.

I'm still not having a working solution yet, I get a low publishing rate (approx 3Hz) so I guess something is wrong with the batcher somehow and I am debugging this at the moment.

My questions are : 
 - What is happening here : https://github.com/ros-drivers/ros2_ouster_drivers/blob/2a2a7db5aebe17ea05b6ad6e04047ad5cfc0f0ad/ros2_ouster/src/ouster_driver.cpp#L208 With the `std::pair` and the iterators. And why to we always take the second to call `process` ? It's a bit fuzzy to me. Is this part of the code documented or could you explain ?
- How is it defined what type of package (imu or lidar) does a processor needs to get ? For example I guess the pointcloud processor should get only lidar packets and not imu. Am I right ? How is it managed ?

Thanks for your help !

SteveMacenski commented:
https://en.cppreference.com/w/cpp/algorithm/equal_range, because the second is the data processor pointer https://github.com/ros-drivers/ros2_ouster_drivers/blob/5f2435dd9443ead4e2ea51786b398e1d8bbf0d65/ros2_ouster/include/ros2_ouster/ouster_driver.hpp#L49 the first is the `ClientState`. 

https://github.com/ros-drivers/ros2_ouster_drivers/blob/2a2a7db5aebe17ea05b6ad6e04047ad5cfc0f0ad/ros2_ouster/include/ros2_ouster/OS1/processor_factories.hpp#L132 the processor factory sets this

MaxandreOgeret commented:
I just got the sensor, kindly lent by Ouster, to allow me to finish this project.

There is something I noticed that I need to figure out now. It's possible now that when asking for the client state, it will return `LIDAR_DATA` AND `IMU_DATA`. 

Here's the output of this piece of code : 
```c++
uint8_t * OS1Sensor::readPacket(const ouster::sensor::client_state & state)
{
  if (state & ouster::sensor::client_state::LIDAR_DATA) {
    std::cout << "LIDAR_DATA" << std::endl;
  }
  if (state & ouster::sensor::client_state::IMU_DATA) {
    std::cout << "IMU_DATA" << std::endl;
  }
  std::cout << "---------------------------------" << std::endl;

[...]

}
```

```
[...]
[ouster_driver-1] ---------------------------------
[ouster_driver-1] LIDAR_DATA
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] IMU_DATA
[ouster_driver-1] ---------------------------------
[ouster_driver-1] LIDAR_DATA
[ouster_driver-1] ---------------------------------
[...]
```

We can clearly see that something strange is happening there, I think that's what is causing my problems at the moment. I'm on it this weekend.

MaxandreOgeret commented:
@SteveMacenski 

After some investigation I think I have noticed a problem that will require some work to be solved.
It seems that running the full driver in a single thread is not enough. At least on my computer. (Ryzen 3900x)

Here's the tests I made with the pointcloud processor : 

||Frequency|
|---|---|
|empty pc (no processing)|`10Hz`|
| `512x10 `|`9.1Hz` And incomplete pointcloud|
| `1024x10`|`4.5Hz` And incomplete pointcloud|
| `1024x20`|`2.8Hz` And incomplete pointcloud|

When publishing empty pointclouds, if I uncomment the `scan_to_cloud` function, then the frequency drops as well. It shows that the processing made in scan_to_cloud is too heavy to run in a single thread along with the rest of the driver. (At least now).

Could it be why Ouster chose to create 2 nodes and hence 2 processes to query the sensor and build/publish the pointclouds ?

How would you test that ?

Thanks for your help !

MOG-Work committed:
handled case with imu data + lidar data, pcl processor working

MOG-Work committed:
imu processor now working

MOG-Work committed:
image processor now working

MOG-Work committed:
scan processor now working + cleaning unused files

MOG-Work committed:
image processor now uses scan instead of cloud

MOG-Work committed:
removed notion of sensor model as f2.0 is universal

MOG-Work committed:
todos and code review

MaxandreOgeret commented:
Should be mostly done by now. Waiting for comments and code-review ! :)

SteveMacenski commented:
I'll review today! What's "mostly done" mean - still some issues to resolve or complete?

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -56,22 +54,37 @@ static const bool IS_BIGENDIAN = []()
+/**
+ * @brief Converts a 4dMatrix into a Vector
+ */
+static std::vector<double> toVector(const Eigen::Matrix<double, 4, 4, Eigen::DontAlign>& mat) {
```
I think this should be `inline` instead

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -181,54 +200,26 @@ inline sensor_msgs::msg::Imu toMsg(
-  sensor_msgs::msg::PointCloud2 msg;
-  pcl_conversions::moveFromPCL(cloud2, msg);
+  sensor_msgs::msg::PointCloud2 msg{};
+  pcl::toROSMsg(cloud, msg);
```
Does this properly add the complete field set from the Point struct? (X,  Y, Z, I, R, t, etc)

MaxandreOgeret replied:
Yes it does, 

Here is the beginning of the pointcloud2 message : 

```
header:
  stamp:
    sec: 1616864489
    nanosec: 122446205
  frame_id: laser_data_frame
height: 128
width: 1024
fields:
- name: x
  offset: 0
  datatype: 7
  count: 1
- name: y
  offset: 4
  datatype: 7
  count: 1
- name: z
  offset: 8
  datatype: 7
  count: 1
- name: intensity
  offset: 16
  datatype: 7
  count: 1
- name: t
  offset: 20
  datatype: 6
  count: 1
- name: reflectivity
  offset: 24
  datatype: 4
  count: 1
- name: ring
  offset: 26
  datatype: 2
  count: 1
- name: ambient
  offset: 28
  datatype: 4
  count: 1
- name: range
  offset: 32
  datatype: 6
  count: 1
```

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -240,42 +231,56 @@ inline sensor_msgs::msg::LaserScan toMsg(
-  }
+  msg.scan_time = 1.0 / ouster::sensor::frequency_of_lidar_mode(mdata.mode);
+  msg.time_increment = 1.0 / ouster::sensor::frequency_of_lidar_mode(mdata.mode) / ouster::sensor::n_cols_of_lidar_mode(mdata.mode);
+  msg.angle_increment = -2 * M_PI / ouster::sensor::n_cols_of_lidar_mode(mdata.mode);
```
Why is this negative?

MaxandreOgeret replied:
Good question, I spent some time on this and it appears that the points are stored in the opposite order in the LidarScan as there vere in version 1.X.
Figured this out with trial and error.

SteveMacenski replied:
ummm ok, I'm not 100% sure this is the right way of handling it. Would it be possible to iterate through it in the opposite direction instead so that this is positive? I'm not sure that all the tools in ROS will handle this correctly.

MaxandreOgeret replied:
Done now

SteveMacenski commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -240,42 +231,56 @@ inline sensor_msgs::msg::LaserScan toMsg(
-  msg.scan_time = 1.0 / rate;
-  msg.time_increment = 1.0 / rate / resolution;
-  msg.angle_increment = 2 * M_PI / resolution;
+  for (size_t i = ls.w * ring_to_use; i < (ls.w * ring_to_use) + ls.w; i++) {
```
So to be clear, this is now buffering the full 3D multi-ring laser data, and this `for` loop is the responsible party for getting just the horizontal beam inputted? 

Just curious - you showed the full 3D pointcloud in a video, can you show the images / 2D laser scan as well to show its working?

MaxandreOgeret replied:
Yes exactly. According to the doc : 

> LidarScan : 
>  * Datastructure for efficient operations on aggregated lidar data.
>  *
>  * Stores each field (range, intensity, etc.) contiguously as a H x W block of
>  * 4-byte unsigned integers, where H is the number of beams and W is the
>  * horizontal resolution (e.g. 512, 1024, 2048).

SteveMacenski commented on ros2_ouster/include/ros2_ouster/client/ouster_ros/point.h:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @file
```
Lets pull this out of the `ouster_ros` directory if its the only file in there

SteveMacenski commented on ros2_ouster/include/ros2_ouster/processors/pointcloud_processor.hpp:
```diff
@@ -41,50 +43,54 @@ class PointcloudProcessor : public ros2_ouster::DataProcessorInterface
+    _batch = new ouster::ScanBatcher(_width, _pf);
+    _xyz_lut = ouster::make_xyz_lut(mdata);
+    _ls = ouster::LidarScan{_width, _height};
+    _cloud = new Cloud{_width, _height};
```
Why not a smart pointer so it handles its own memory deallocation?

MaxandreOgeret replied:
Make sense, I used new because in `processor_factories.hpp` smart pointers were not used.

For example here : 

```c++
inline ros2_ouster::DataProcessorInterface * createIMUProcessor(
  const rclcpp_lifecycle::LifecycleNode::SharedPtr node,
  const ouster::sensor::sensor_info & mdata,
  const std::string & frame,
  const rclcpp::QoS & qos,
  const ouster::sensor::packet_format& pf)
{
  return new sensor::IMUProcessor(node, mdata, frame, qos, pf);
}
```

Now  `processor_factories.hpp` and the processors all use smart pointers.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/processors/pointcloud_processor.hpp:
```diff
@@ -41,50 +43,54 @@ class PointcloudProcessor : public ros2_ouster::DataProcessorInterface
-      std::make_shared<pcl::PointCloud<point_os::PointOS>>(_width, _height);
+    _height = mdata.format.pixels_per_column;
+    _width = mdata.format.columns_per_frame;
+    _batch = new ouster::ScanBatcher(_width, _pf);
```
Same with this - why not a smart pointer?

SteveMacenski commented on ros2_ouster/include/ros2_ouster/processors/scan_processor.hpp:
```diff
@@ -26,44 +26,43 @@
+    _height = mdata.format.pixels_per_column;
+    _width = mdata.format.columns_per_frame;
+    _xyz_lut = ouster::make_xyz_lut(mdata);
+    _batch = new ouster::ScanBatcher(_width, _pf);
```
Ditto on smart pointers

SteveMacenski commented on ros2_ouster/params/sensor.yaml:
```diff
@@ -49,4 +49,4 @@ ouster_driver:
     # PCL
     #
-    os1_proc_mask: IMG|PCL|IMU|SCAN
+    proc_mask: IMG|PCL|IMU|SCAN
```
Add new line at end of file

SteveMacenski commented on ros2_ouster/include/ros2_ouster/processors/image_processor.hpp:
```diff
@@ -0,0 +1,235 @@
+    _width = mdata.format.columns_per_frame;
+    _px_offset = mdata.format.pixel_shift_by_row;
+    _ls = ouster::LidarScan{_width, _height};
+    _batch = new ouster::ScanBatcher(_width, _pf);
```
Smart pointers?

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -48,7 +48,7 @@ OusterDriver::OusterDriver(
   this->declare_parameter("imu_frame", std::string("imu_data_frame"));
   this->declare_parameter("use_system_default_qos", false);
-  this->declare_parameter("os1_proc_mask", std::string("IMG|PCL|IMU|SCAN"));
+  this->declare_parameter("proc_mask", std::string("IMG|PCL|IMU|SCAN"));
```
Did you change this variable name in the config files?

MaxandreOgeret replied:
Yes, it works now

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -197,24 +195,36 @@ void OusterDriver::broadcastStaticTransforms(
       ros2_ouster::toString(state).c_str());
 
-    uint8_t * packet_data = _sensor->readPacket(state);
+    _lidar_packet_data = _sensor->readLidarPacket(state);
```
Its not immediately clear to me why this couldn't be done in a single function - what triggered this change?

MaxandreOgeret replied:
I created two functions so we can get a lidar package and a imu package at once.
I posted above that it is possible to have both imu and lidar data available.

Now I make sure that it's possible to get both

SteveMacenski commented on ros2_ouster/src/sensor.cpp:
```diff
@@ -0,0 +1,129 @@
+// Copyright 2020, Steve Macenski
```
Only last thing is to run `ament_uncrustify --reformat` (auto-fixes) and `ament_cpplint` (needs manual fixing) to get this up the static analysis specs for ROS2.

MaxandreOgeret replied:
I uncrustified and pushed :)

SteveMacenski reviewed (changes requested):
Very good overall - assuming this is fully functional for firmware V2, this is very promising looking with just minor updates.

I'd like to know though if there's a mechanism (and if it works) to reject on start up / throw an error / something for if you try to use this with an incompatible firmware version (e.g. 1.X) to let someone know they need to update firmware to function properly.

MaxandreOgeret commented:
Hi and thanks for your time !

>  What's "mostly done" mean - still some issues to resolve or complete?

It should mean everything except the timestamp. I haven't tested them and I plan to spend this Saturday on this and on your review.

> which files, if any, you took directly from the Ouster Example repo without changes.

All the files under the `client` dir in src and include : 
- `ros2_ouster_drivers/ros2_ouster/include/ros2_ouster/client/`
- `ros2_ouster_drivers/ros2_ouster/src/client/`

> anything you want me to specifically comment on from above?

No I think it's good and that those problems are resolved.

> Very good overall

Thanks ! I spent a lot of time on this and it's greatly appreciated ! I tried to write clean code but I know it's not perfect and it's priceless for me to get feedback on the code I write. Thanks a lot !

> I'd like to know though if there's a mechanism (and if it works) to reject on start up / throw an error / something for if you try to use this with an incompatible firmware version (e.g. 1.X)

Good question, I know it's possible to query the firmware version, but I dont know how how the version 1.X would react if we try with the @.0 client. 
Needs testing. Maybe we can check using the TCP API and then start the driver if the version is 2.0 ?

SteveMacenski commented:
I think its OK if we connect to the  sensor to get its firmware, but at the first opportunity, it would be great to throw errors galore or throw an exception to bring everything down if in fact the firmware is incompatible with this driver version (requiring 2.x or whatever the full valid range of this client is)

MaxandreOgeret commented:
I would need a firmware 1.X sensor to test that. Making an API call seems doable to identify the firmware. I would have to look at the software documentation of the firmware 1.X to see if we can use the same API call as firmware 2.0.

Here's the scan working :  

- Scan + pointcloud
  ![Screenshot from 2021-03-26 17-52-07](https://user-images.githubusercontent.com/22786578/112720174-6e1dfe80-8f05-11eb-9e59-50ce2109087e.png)

- Scan only
  ![Screenshot from 2021-03-26 17-52-02](https://user-images.githubusercontent.com/22786578/112720183-81c96500-8f05-11eb-844c-2cb305b3fde0.png)

- Pointcloud only
  ![Screenshot from 2021-03-26 17-52-09](https://user-images.githubusercontent.com/22786578/112720193-9279db00-8f05-11eb-874d-d07f62017512.png)

MaxandreOgeret commented:
About the timestamp I have decided to make a change on how we enable the timestamp to be overridden by the ROS time. And I think using the Ouster timestamp_mode for this matter is a bad idea.

Right now you can user the ros time by setting the timestamp mode to `TIME_FROM_ROS_RECEPTION` which doesn't exist ion the sensor. And at the moment if I try to set this timestamp I will get the following error : 

> Invalid timestamp mode: TIME_FROM_ROS_RECEPTION

The solution would be to edit the Ouster client to accept it and send another type of timestamp mode that is valid. But it wouldn't represent what is truly happening within the sensor.

Also I think we should avoid to edit the ouster client at all cost to make upgrading easier later. 

I will then create a new boolean parameter : `use_ros_time` with an explanation in the config yaml 

Another solution would to send `TIME_FROM_INTERNAL_OSC` to the sensor if the timestamp mode in the driver is set to `TIME_FROM_ROS_RECEPTION`.

What do you think ? I think it makes more sense that way. If you think it's bad idea we can find something else but I really believe we shouldnt change the ouster client at all.

Thanks !

MOG-Work committed:
removed const for packetformat in the processors as it would prevent to access its accessors

MOG-Work committed:
use ros time now works, made data const in processors

MOG-Work committed:
use smart pointers in processors

MOG-Work committed:
processor factories now use smart pointers. Added EOL at the end of conf yaml file

MOG-Work committed:
uncrustified

MaxandreOgeret commented:
I'm quite happy with how it goes now. I just have one thing that bothers me.
 If you look in those processors : 

- `scan_processor.hpp`
- `pointcloud_processor.hpp`
- `image_processor.hpp`

You can see that we batch the packets in all of them. It seems to me that this is very redundant as it will batch 3 times erxactly the same thing. I think we should batch only once and all those processors should just use the generated `LidarScan`.

How would you do that ? Should we create pre-processors that the processors can access ?

Thanks ! :)

SteveMacenski commented:
> I would need a firmware 1.X sensor to test that.

True, but I'm just asking if there's anything in place to do that in this PR, even if it hasn't been tested yet

Scan stuff looks good! Thanks for showing

I don't understand the time concerns, it was working before, why mess with it? The whole point of the timemodes is so that there **arent** just a bunch of boolean params for which time mode to use, there's just the one `timestamp_mode` that @tpanzarella wrote. 

> You can see that we batch the packets in all of them. It seems to me that this is very redundant as it will batch 3 times erxactly the same thing. I think we should batch only once and all those processors should just use the generated LidarScan.

Before this V2, the V1 software had lambdas that would take in raw packets, assemble them, and then publish on a full rotation. It looks like on V2 they support some kind of assembler for us. If there's a clean way for a global batcher to notify us that 'hey, I have a full scan' that we can then use to all the appropriate processors to populate the appropriate ROS types / data, I'd be OK with that.

MaxandreOgeret commented:
> True, but I'm just asking if there's anything in place to do that in this PR, even if it hasn't been tested yet

There is a way to get the sensor version, but it happens after the `init_client` and I think it would fail with a 1.X sensor.

> I don't understand the time concerns, it was working before, why mess with it? 

Because it didn't work anymore, I didn't 'mess' with it for no reason.  [This](https://github.com/ros-drivers/ros2_ouster_drivers/blob/6434c7c016adbb7cb555718bf12c60489aac93a7/ros2_ouster/src/sensor.cpp#L48) test would fail as `ouster::sensor::timestamp_mode_of_string` would report an unknown time mode. So something had to be done.

> Are you just trying to get the sensor to work properly with this different timestamp mode (e.g. nothing changes how the timestamps are assigned, but the sensor is just unaware of it)?

Yes exactly, and I detailed the two ways of doing that above. I will revert the changes I made, and use `timestamp_mode` instead.

> It looks like on V2 they support some kind of assembler for us

Yes it's the [scanbatcher](https://github.com/ros-drivers/ros2_ouster_drivers/blob/6434c7c016adbb7cb555718bf12c60489aac93a7/ros2_ouster/src/client/lidar_scan.cpp#L71). I can imagine a batcher pre-processor that the processors would poll and eventually get a Lidarscan once it's ready. 

What do you think ? Any implementation ideas ?

SteveMacenski commented:
> There is a way to get the sensor version, but it happens after the init_client and I think it would fail with a 1.X sensor.

I'd still just check, if nothing else it creates the check for later development to check specific version if required. 

Your comments on timestamp make sense, but just verifying that you're still using the ROS time in that situation, regardless of what the actual lidar sends back in this case? 

Without reviewing the nitty-gritty specifics of the scan batcher, I assume there's some way that the processors today know when the complete data is received in order to populate and send out the ROS messages? If so, then you could move the batcher up one level of abstraction and check if its complete there. If so, then call the `N` processors in the processor map that depend on the laser data. 

The one issue here is if someone wants to actually do something live with the data rather than waiting for the full scan. I envisioned a future desire to potentially batch in smaller chunks (like 10 packets) for low latency incomplete pointcloud sending to help de-alias while operating in high speed environments. This would stop that from working. Letting each processor handle everything themselves leaves that flexibility for new processors to do specific things with the packets rather than only after getting all of the data batched up in a full rotation.

MOG-Work committed:
scan angle increment now positive

MaxandreOgeret commented:
- Now using the `timestamp_mode` parameter to set time from ROS. It will initialize the sensor with `TIME_FROM_INTERNAL_OSC` and stamp the messages with ROS time.
- The scan `angle_increment` is now positive. I iterate through the lidarscan in the reverse direction.

MOG-Work committed:
better access for TF eigen mat

MOG-Work committed:
removed useless cloud conversion in scan processor

SteveMacenski commented on ros2_ouster/include/ros2_ouster/preprocessors/preprocessor_manager.hpp:
```diff
@@ -0,0 +1,79 @@
+ * @brief This class manages the preprocessors. It will create the preprocessors
+ * It is also passed to the processors so thery can access the preprocessors
+ */
+class PreprocessorManager
```
This object seems overkill - why not just use the preprocessor directly? This simply just calls its API of the same name and its not likely this will ever be expanded to include more preprocessors

MaxandreOgeret replied:
The only reason for this object is to give the processors an access point to the preprocessors. I thought that if processors have an interface, preprocessors should have on too. But if you think we will never create any other preprocessor then yes it's useless.

SteveMacenski replied:
Lets remove the concept of "preprocessor" and just use the `FullRotationAccumulator`/`LidarscanPreprocessor` object here

MaxandreOgeret replied:
How would you suggest the implementation of the FullRotationAccumulator ?
- Remove preprocessor_manager.hpp 
- Rename lidarscan_preprocessor to FullRotationAccumulator ?

And how would you deactivate this FullRotationAccumulator if no activated processor needs it ?

SteveMacenski replied:
Yes, exactly that :-) 

You wouldn't just assume that someone needs it (in practice, everything needs it)

MaxandreOgeret replied:
I see, it' s seems reasonable. 
I agree, even though it's possible to use this driver only for the imu

SteveMacenski replied:
If someone used this just for the IMU, I think then we can cross that bridge, but that's a very expensive and power hungry IMU to use an Ouster lidar :wink:

MaxandreOgeret replied:
Done

SteveMacenski commented on ros2_ouster/include/ros2_ouster/preprocessors/lidarscan_preprocessor.hpp:
```diff
@@ -0,0 +1,132 @@
+   */
+  bool isDataReady() override
+  {
+    _activated = true;
```
Why would asking if data is ready activate it? I don't think this activate variable does anything useful

MaxandreOgeret replied:
Because I imagined the situation where we have several preprocessors and several processors. And each processor needs between zero and several preprocessors. And not all processors are activated with the mask so some preprocessors are not needed.

With this solution if a processors checks for data it will activate the preprocessor. So all the preprocessors that are not used will not process any data.
It seemed a good solution because the activated preprocessors that need a specific preprocessor will just activate it, and it requires no extra logic.

Let's say only the imu processor is activated. If we remove this variable we need to add extra logic to disable the lidarscan preprocessor. With this solution it's totally transparent.

MaxandreOgeret replied:
Removed this logic.

SteveMacenski reviewed (commented):
This preprocessor interface / manager / etc I think is getting way overcomplicated - why not just use the LidarscanPreprocessor directly? That seems good enough and remove these other levels of abstraction. Lets cross the other sensor batching bridge when we come to it - I know I wanted to support it, but I don't want to over-engineer a solution, a processor could always ignore the full rotation accumulation tool and accumulate itself locally for whatever batch size it likes if it doesn't want to use that one as an input tool. Practically speaking, this full rotation accumulation tool will never be replaced but the processors locally might buffer in different sizes in that situation.  How do you feel about the name `FullRotationAccumulator` to replace LidarscanPreprocessor? I think its more descriptive and so far all of our processors just it.

MaxandreOgeret commented:
@SteveMacenski Your last remarks have been solved. 

Waiting for your feedback !

SteveMacenski commented on ros2_ouster/include/ros2_ouster/full_rotation_accumulator.hpp:
```diff
@@ -0,0 +1,122 @@
+  std::shared_ptr<ouster::LidarScan> getLidarScan()
+  {
+    if (!_batchReady) {
+      throw ros2_ouster::OusterDriverException("Full rotation not accumulated.");
```
Are these 2 get functions wrapped in a try/catch in case this throws?

MaxandreOgeret replied:
Yes, they will trigger this try/catch.
 https://github.com/ros-drivers/ros2_ouster_drivers/blob/ebded45964a46e1e60f219a75deb1d1e1ca89457/ros2_ouster/src/ouster_driver.cpp#L201

SteveMacenski commented on ros2_ouster/include/ros2_ouster/full_rotation_accumulator.hpp:
```diff
@@ -0,0 +1,122 @@
+  bool accumulate(const uint8_t * data, uint64_t override_ts)
+  {
+    if (_batchReady) {
+      _batchReady = false;
```
What happens if no one polls for a new batched scan before the next `accumulate()` is called? Would we just drop a measurement?

SteveMacenski replied:
In context would this be possible to happen? 

Maybe we should use that return type `bool` to be `true` if ready with a new full rotation and `false` if not. Even if we don't use it right now, it would make sense for that to be the return type here -- if its done accumulating

MaxandreOgeret replied:
> What happens if no one polls for a new batched scan before the next accumulate() is called? Would we just drop a measurement?

Yes we would, but we would call `accumulate` only when there is a `lidar_packet` available to build a new batch.

> Maybe we should use that return type bool to be true if ready with a new full rotation and false if not. Even if we don't use it right now, it would make sense for that to be the return type here -- if its done accumulating

The problem is that we need to accumulate data before the processors, in `ouster_driver.cpp`, here : 

https://github.com/ros-drivers/ros2_ouster_drivers/blob/ebded45964a46e1e60f219a75deb1d1e1ca89457/ros2_ouster/src/ouster_driver.cpp#L217

That's the whole point of the accumulator. The processors don't call `accumulate()` So what would we do with this return value in `ouster_driver.cpp`?

`accumulate()` could return `_batchReady` but then for what ? This return value would not be used anywhere. And the processors cannot call `accumulate()` because it would then accumulate with the same packetdata that was already accumulated in `ouster_driver.cpp`.

We can make `accumulate()` return type void though.

SteveMacenski replied:
>    What happens if no one polls for a new batched scan before the next accumulate() is called? Would we just drop a measurement?

I believe in context of the code the accumulate is called and directly after the processors are called checking batched scans, correct? So there shouldn't be a situation where we're messing that up, correct?

> The problem is that we need to accumulate data before the processors, in ouster_driver.cpp, here :

Imagine this:

```
bool scan_ready = _full_rotation_accumulator->accumulate(_lidar_packet_data, override_ts); 
if (scan_ready) {
  RCLCPP_DEBUG(logger_, "Full rotation scan is ready for processing...");
}
```

Even if it doesn't have alot of use (right now) a future refactor could make use of that return code to do something interesting (for instance, someone using this to separate out full-scan processors from packet processors and only call full-scan ones if true). Right now you return a bool that is always true. I suppose my suggestion is to return `void` if we're not going to have that bool be meaningful, or return `true/false` based on if the scan is fully accumulated so that it does mean something.

So yeah, I don't really care too much, but if we're only going to return `true`, may as well be `void`. That debug log might be helpful though if in the future we have performance issues to see the regularity of that rate.  That's a pretty weak argument for it, but that's what I got :laughing:

MaxandreOgeret replied:
>I believe in context of the code the accumulate is called and directly after the processors are called checking batched scans, correct? So there shouldn't be a situation where we're messing that up, correct?

That is correct, and I don't see any situation where it could mess up.

> for instance, someone using this to separate out full-scan processors from packet processors and only call full-scan ones if true

That is true, but wouldn't it be more explicit to call `FullRotationAccumulator::isBatchReady()` directly ? So we would get something like this : 

```cpp
_full_rotation_accumulator->accumulate(_lidar_packet_data, override_ts); 
bool scan_ready = _full_rotation_accumulator->isBatchReady(); 
if (scan_ready) {
  RCLCPP_DEBUG(logger_, "Full rotation scan is ready for processing...");
}
```
It's one extra line but checking for the readiness of the batch would be consistent in all the source. In that case it would make sense for `accumulate()` to return void.

> Right now you return a bool that is always true.

I agree that it doesn't make sense to always return true.

> That debug log might be helpful though if in the future we have performance issues to see the regularity of that rate.

It would be a good enough reason for me ! But it also be possible to use `FullRotationAccumulator::isBatchReady()` for that ! :laughing:

SteveMacenski replied:
It could also be possible, yes. I don't really care too much, change to `void` or have the `bool` mean something other than always `true` -- dealers choice

MaxandreOgeret replied:
Done !

SteveMacenski commented on ros2_ouster/params/sensor.yaml:
```diff
@@ -1,7 +1,7 @@
   ros__parameters:
-    lidar_ip: 10.5.5.96
-    computer_ip: 10.5.5.1
+    lidar_ip: 169.254.140.51
```
Revert these please :-) I think these are made to match the readme remarks on setup

MaxandreOgeret replied:
oops That's a mistake

SteveMacenski commented on ros2_ouster/include/ros2_ouster/processors/scan_processor.hpp:
```diff
@@ -88,18 +82,15 @@ class ScanProcessor : public ros2_ouster::DataProcessorInterface
    */
   void handler(const uint8_t * data, const uint64_t override_ts)
```
Can these handler functions in the processors be put back into `process()`? I think they all just call `handler()` anyhow with the same signatures. Its a little odd to me its split off

MaxandreOgeret replied:
The main reason to split off here is to have public assessors that eventually call private processing functions. Like I did here :

https://github.com/MaxandreOgeret/ros2_ouster_drivers/blob/foxy_devel_f2.0/ros2_ouster/include/ros2_ouster/full_rotation_accumulator.hpp 

But I agree that it's not necessary and I will put order into that :)

SteveMacenski replied:
I'm not sure why you'd make a private function to do that vs just in the public function itself? This could be a totally legit design method - I just haven't run across it before, can you tell me a little more about that choice?

MaxandreOgeret replied:
First of all i think it makes the code readable as it allows to have the processing function only focused on processing and the assessor focused on other things like right management or whatever.
Also it's because of some frameworks I worked with that were Service-oriented. The private function could easily be exported to a new service.

MaxandreOgeret replied:
Done !

SteveMacenski reviewed (commented):
Looks very good to me, just a couple of questions mostly

MOG-Work committed:
added full rotation accumulator + decrustified

MOG-Work committed:
removed useless handler function in accumulator and processors

MaxandreOgeret commented:
@SteveMacenski Your last remarks have been solved.

Waiting for your feedback !

SteveMacenski reviewed (approved):
This, now, looks done to me! Next steps:
- Find someone with a V2 firmware sensor to sanity check us (cpu is still reasonable, getting full framerate, topics are publishing, etc) 
- Figure way to document / branch this should go in to make it clear to people this requires firmware version V2 (maybe a log in caps on startup?) as matched to the distro release. We may also want to log on startup the firmware version of the sensor its connected to (in case they just don't actually know). This can't go itself into `foxy-devel` since its been released and this would break everyone currently using it, but we could do a release under a slightly different name `ouster_drivers_v2` for foxy instead if someone wanted V2 firmware compatible binaries. 
- Determine if we want to release this to Foxy under a V2 name or just for future distros (Galactic soon!)

MaxandreOgeret commented:
How will we proceed ? It would be sad for the PR to be stuck at this point.

For the version compatibility a log on startup is the easiest thing to do., also it's very easy to log the sensor version once the driver is connected to it.

I think it would be nice to release it for foxy still, I believe it has a lot of users still.

MaxandreOgeret commented:
Found a bug that is also present in the current Ouster client.

On this line : https://github.com/MaxandreOgeret/ros2_ouster_drivers/blob/114406409ce40415e3b05fd1d2c4e75b0d6faa71/ros2_ouster/include/ros2_ouster/conversions.hpp#L268 We should set the `RANGE` and not `REFLECTIVITY`.

According to the definition of the Point : https://github.com/MaxandreOgeret/ros2_ouster_drivers/blob/foxy_devel_f2.0/ros2_ouster/include/ros2_ouster/client/point.h

I will correct that very soon.

MaxandreOgeret commented on ros2_ouster/include/ros2_ouster/conversions.hpp:
```diff
@@ -181,99 +187,87 @@ inline sensor_msgs::msg::Imu toMsg(
+        static_cast<uint16_t>(pix(ouster::LidarScan::REFLECTIVITY)),
+        static_cast<uint8_t>(u),
+        static_cast<uint16_t>(pix(ouster::LidarScan::AMBIENT)),
+        static_cast<uint32_t>(pix(ouster::LidarScan::REFLECTIVITY))};
```
Found a bug at this line that is also present in the current Ouster client. We should set the `RANGE` and not `REFLECTIVITY` here.

According to the definition of the Point : https://github.com/MaxandreOgeret/ros2_ouster_drivers/blob/foxy_devel_f2.0/ros2_ouster/include/ros2_ouster/client/point.h

I will correct that very soon.

MaxandreOgeret replied:
Solved.

MaxandreOgeret committed:
Corrected bug where REFLECTIVITY was put in place of RANGE in the Point.

SteveMacenski commented:
I created a new ROS2 branch, lets retarget this PR to that so that it will be used as the basis of the next distribution release. 

> For the version compatibility a log on startup is the easiest thing to do., also it's very easy to log the sensor version once the driver is connected to it.

Lets add that here. Also, it would be good if the README in the first section said that it requires V2.x firmware to work, that way you can save alot of people problems and headaches that they know off the bat this version requires it. 

> I think it would be nice to release it for foxy still, I believe it has a lot of users still.

Its not ABI compatible since it requires a hard firmware change, this couldn't be released to foxy under the same name. We could release it for the foxy distribution, but it would have to be under a new package name on a different branch (e.g. ouster_drivers_v2)

The steps I think are pretty clear
- Get someone else with a sensor to sanity check us that this works for them
- Merge into `ros2`, once Galactic is released, release this to galactic
- Decide if we want to release also to Foxy for a v2 version to use

anderwm commented:
Sorry, I was re-vectored onto another effort so I haven't been following too closely. I will try to give it a look.

MaxandreOgeret commented:
@SteveMacenski I have changed the target of this PR to the branch `ros2`.

anderwm commented:
It is definitely working. ![Screenshot from 2021-04-23 07-44-51](https://user-images.githubusercontent.com/1276785/115873157-3c1a9200-a408-11eb-89d7-8af19fb3c857.png)

I don't have time at the moment to investigate the more complicated pieces I will ultimately need (using as component). I am using an OS0-128-U and I also haven't investigated if it is using all the scans/being used to it's full potential. However I noticed a couple of things (Not saying it's a problem yet, just that I noticed it).

1. I have to set the Reliability Policy to Best Effort in RViz to view anything
2. Only the range_image renders in RViz. ambient_image and intensity_image get dropped for not have a FrameID [INFO] [1619182373.389459026] [rviz]: Message Filter dropping message: frame '' at time 649.826 for reason 'EmptyFrameID'

If there is anything specific you worry about/want tested shoot me a message, otherwise thanks for you work on this.

SteveMacenski commented:
1. That's just a ROS2 thing, sensors should be best effort by default
2. That's a problem @MaxandreOgeret did it not populate the frames for those? 

But overall, looks great!

MOG-Work committed:
added frame_id for ambient_image+intensity_image. Added log about fw requirements. Added readme line about fw requirements

MaxandreOgeret commented:
Hey, sorry for the late replay ! I ĺl be more available from now :)

I didn't get this error with the images :

![image](https://user-images.githubusercontent.com/22786578/116822665-b5448280-ab88-11eb-86de-2ec0085cf124.png)

But indeed the frame ID was only set for the range image. I corrected that.

I also added the message about the firmware requirements in the readme : https://github.com/ros-drivers/ros2_ouster_drivers/blob/1dba63270ec609f4b51b082c0e466418cdeab1a9/README.md#L3

And the log : https://github.com/ros-drivers/ros2_ouster_drivers/blob/1dba63270ec609f4b51b082c0e466418cdeab1a9/ros2_ouster/src/ouster_driver.cpp#L61

SteveMacenski commented:
Awesome, if @anderwm or someone can verify that, I'm happy to merge

anderwm commented:
![Screenshot from 2021-05-03 13-15-59](https://user-images.githubusercontent.com/1276785/116917168-eae27d80-ac13-11eb-9f0f-e4b240664026.png)
![hz_data](https://user-images.githubusercontent.com/1276785/116917377-3006af80-ac14-11eb-8390-97179f1948e6.png)

Better. I'm not sure how to verify it is using all 128 scans, but looks good so far.

SteveMacenski commented:
Awesome! @MaxandreOgeret anything further before we merge this in?

MaxandreOgeret commented:
Nothing to add at the moment ! 
Thanks a lot @SteveMacenski for your advice and your guidance. Thanks @anderwm for your test !
If you have other interesting projects that need some coding it's a pleasure !

SteveMacenski commented:
Merged, `ros2` now set to the default branch, and will be released with Galactic! 

I spend by far the largest chunks of my time on Nav2 in ROS2 https://github.com/ros-planning/navigation2, where there are a ton of different projects to be worked on! We also have a community slack (https://join.slack.com/t/navigation2/shared_invite/zt-pv43supx-hfN_Q3RfaKkn9f8i3Wt~OQ) that might be easy to discuss through as well. I'd be more than happy to have you involved on something else!!
